### ISSUE TITLE
Ethos 1.5.11 Model Send / Receive

### ISSUE THREAD

**Issue by BJR981S** [FrSkyRC/ETHOS-Feedback-Community#4343]
Updated.

When you send a model to a different radio it does not transfer all the special functions that are configured in the model. It only transfers non global special functions. Any Special function marked as Global are not transferred.

To minimise the impact on the destination radio, the Global Special Functions need to be deactivated as Global and become Local to the model being sent.

This will minimise the number of Global Special functions listed on the destination radio if multiple models are being sent. 

This makes life very difficult for those of us with Multiple ethos radios. And those of us that want to test Ethos versions on a test radio without having to recreate multiple models for testing.

**Comment by pmullen503** [comment]
I would not want it to send global functions. That would mean when I got a file from a buddy to look at, that would essentially reset all MY global functions to his.

**Comment by BJR981S** [comment]
> I would not want it to send global functions. That would mean when I got a file from a buddy to look at, that would essentially reset all MY global functions to his.

No thats not correct. It would add his Special Function Global functions to yours. When Special Functions are copied it allocates a new number to the function. It does not overwrite. It renumbers and reallocates the new numbers to the copied model. Try copying a model with Special functions. Not Global and you will see what I mean about the renumbering.

**Comment by pmullen503** [comment]
I perhaps shouldn't have used the term reset, a better word would have been contaminate.

From the manual: "When selecting Global, the special function is added to all existing models and any new model created in the future."

So bringing over someone else's global functions changes all of your models, past and future.

**Comment by BJR981S** [comment]
> I perhaps shouldn't have used the term reset, a better word would have been contaminate.
> 
> From the manual: "When selecting Global, the special function is added to all existing models and any new model created in the future."
> 
> So bringing over someone else's global functions changes all of your models, past and future.

Not really. It does not change them. Global functions are a radio configuration not a model configuration. It would make those new global function available to older models but you would need to specifically add one or more to the old model to impact the model configuration. New models would only be linked to the additional Global function if you specifically added it to the model as a setting as well.

The only downside is that your Global Functions list would expand to be made available to all model configurations on that radio.

**_May I suggest then that if this is implemented to prevent contamination of your available global functions by addition (making the available function list longer) that the send and receive model function resets the global setting of the function as part of the Send Process. Would that be acceptable to you?_**. That would have no impact on anything as non global functions are attached to the model config anyway. As an Instructor copying models across radios is quite painful as it is. I have 17 Ethos radios and it's hard enough just managing the FW updates across them. 

The only issue I see with this is that copying a specific Special Function with a function that is not ubiquitous across all Ethos models would be an issue. But that exists for all Special Functions that are sent with a model send, even non global ones.

Can you raise an issue that the manual is incorrect as you have quoted. it is not "**added to all existing models**". It is "**made available**" to all existing models.  You must allocate the function to the model. The wording is most misleading. Can you reference the page in the manual? i.e. You will see all the global functions from that models configuration as well as any non global functions. But to actually use a special function you must create a request to utilise the functions that are available. You can look at it the same way as you look at the specific mixes available on the radio. The mixes are always available but you need to use them to be made a part of the model configuration.

Let me know your thoughts on the suggestion I listed above.

**Comment by pmullen503** [comment]
I just created a global function one one model (Ethos 1.5.11) and it is added and **active** on all models.  If I delete or inactivate it on one model, it's deleted or inactivated on all models.

In this case, I knew it was there and what it did so I could delete it.  Imagine a situation where you had a special function that did something and downloaded a model which had a global function that did the opposite, like involved in throttle cut. Both functions would be active, cancelling each other out and unless you knew the downloaded model had that global special function and could find it, you'd have to figure out why ALL your models are now screwed up.

I would not want global functions transferred with models.

**Comment by BJR981S** [comment]
I'm sorry, this is getting verbose.

1) I have never said that on the model send that the Special Functions had to remain as Global. Go back and Check. 
2) You raised a misunderstanding of how and What Global functions are: I tried to be helpful and explain to you. Im sorry I failed. It has then just confused this issue thread.
2) You are confusing the difference between available and active. I tried to explain it to you. I even gave you an analogy with the fixed and free mixers..
3) You state in your test that they were available and active. You misunderstand what active is. Maybe it's purely semantics. Special Functions that are available have to be activated by a trigger. All special functions even if not global are created and remain "AVAILABLE"  with the model config THEY MUST BE TRIGGERED by logic in the models functions. _**Seeing them in a list is not activation**_. Adding and deleting from a list is **_not activation and deactivation._**
5) I asked you to comment on my suggestion (it was meant to get you to understand the difference between what I have asked for as opposed to discussing how Global Special Functions work.  The suggestion I highlighted and asked for your concurrence in that they did not have to be sent as global, just sent. 

**My last attempt at explaining Special Functions Local And Global.**

**Local Special Functions** are linked to the Model configuration. They are Linked (So can only been seen in the special Functions List on that model only) but not activated. Activation requires a trigger.
**Global Special Functions** are a radio based functions and thus span all Models on the radio (So they can be seen in the Special Functions list on that and All Models) and are available to be activated by a model trigger.

Basically the list of Special Functions in the Special functions is just that a List of what you can trigger. It's a Use these if you want to. The list is an aggregation of Global and Local Special Functions. Just that, a list of functions available to use.

I give you this to ponder.

You are using a radio and have 50+ models on it. You have a couple of new models and you don't want to program the same Special Function across the new models. So you create a new Special Function and make it Global. 

Does it go back and break all your previous 50 models. No It does not. But if you edit those old models you will see the Global Special Function are now in the available Special function List and can be triggered if you so desire.

The logic you are describing, means that Global Special functions can only be used when the radio is virgin, prior to any models being created, and once the first model is created you can no longer create global Special Functions. 

Do you think that is what the architects of ethos had in mind?

I am going to go back and suggest to the developers how to execute this request. I wanted to leave it to them how they should implement, but now I think I will need to be specific.

**Comment by BJR981S** [comment]
> I just created a global function one one model (Ethos 1.5.11) and it is added and **active** on all models. If I delete or inactivate it on one model, it's deleted or inactivated on all models.
> 
> In this case, I knew it was there and what it did so I could delete it. Imagine a situation where you had a special function that did something and downloaded a model which had a global function that did the opposite, like involved in throttle cut. Both functions would be active, cancelling each other out and unless you knew the downloaded model had that global special function and could find it, you'd have to figure out why ALL your models are now screwed up.
> 
> I would not want global functions transferred with models.

The example you articulate is not valid. Additional Global Special Functions are only in an availability list. They are not activated. The identity of the Special Function has a human recognisable name it means nothing. The actual special function I believe is based on a GUID identity. Only triggered Special functions do anything. They are benign while they sit in a list and not triggered by a logical function.

Your misunderstanding is that when you see them in the Special Function List they do something. An easy way for you to test is to create some special functions. They can be global or local. In that list they will be greyed out.  Create a logic switch that triggers one of the special the Special Functions you created. You will see that the Special function will become highlighted when you activate the switch. This is the Special Function being activated. Special functions that do not have activation logic do nothing other than appear in the list.

The one that you have activated with the switch. Copy / clone it. It will look the same but the first will continue to change with the switch, the second will not as it needs specific logic to activate it. It looks the same but has a specific different identity.
